### DISCUSSION TITLE
Getting the value of a data block

### DISCUSSION THREAD

**Discussion by chiffreKing** [NationalSecurityAgency/ghidra#3655] [Q&A]
Hello guys,

I'm having troutble to get to the value of a constant of a DAT object.

<img src="https://user-images.githubusercontent.com/87306782/142619338-8cb00416-ec6b-4393-a021-7fc9f4755c68.png" width="300">

What I'm doing so far is to iterate through the instructions and filter for the 'ldr' mnemonic. Then I'm accessing the opObject(1) and am getting the address of the DAT object.

`for func in funcs:
    instr = listing.getInstructionAt(func.getBody().getMinAddress())
    
    #iterate through every function from its min to max address
    while instr.getMinAddress() <= func.getBody().getMaxAddress():
                
            mnemonic = instr.getMnemonicString()
            argument = instr.getOpObjects(1)[0]
            
            if mnemonic == 'ldr':
                opType = instr.getOperandType(1)
                opTypeScalar = 24576
                if opType == opTypeScalar:
                    print(instr)
                    value = argument.toString()` 

until here it works fine.  but from here on I don't know what to do. I tried 
`newInst = currentProgram.getListing().getInstructionAt(addr(value))`
to get to this address but I get this error:

NameError: name 'addr' is not defined

What I want to do is to get to the address of the DAT object and then read out the string '07000011h'

Please help me out. Thanks

**Comment by chiffreKing** [comment]
I just figuered out the problem. The problem is that I'm trying to access data with an instruction what won't work. What I do now is the following:
`
        mnemonic = instr.getMnemonicString()
        argument = instr.getOpObjects(1)[0]
        
        if mnemonic == 'ldr':
            opType = instr.getOperandType(1)
            opTypeScalar = 24576
            if opType == opTypeScalar:
                print(instr)
                value = argument.toString()
                print(value)
                dataAddress = parseAddress(value)
                dataBlock = getDataAt(dataAddress)
                datConstant = dataBlock.getValue()
                print(datConstant)
                instr = instr.getNext()
`

I'm still asking myself if this is the clean way?

**Reply by astrelsky** [reply]
You *should* be able to use `CodeUnit.getPrimaryReference(1)` on `ldr` to get a `Reference` object and then using `Reference.getToAddress` to get the address of the data. Parsing an address from a number or string is ill-advised due to how Ghidra handles address spaces.

It has been a while though so `getPrimaryReference` may not be the correct method. If it doesn't work try one of the other reference related methods in the `CodeUnit` interface.

**Reply by chiffreKing** [reply]
I tried using `CodeUnit.getPrimaryReference(1)` and `Reference.getToAddress` but it didn't work due to the fact that the obejct of `Reference` is not a member of `ghidra.program.model.address.Address` . so I tried `CodeUnit.getAddress(1)` and that worked just fine.

Thank you very much I really appreciate your support and the general support here.

**Reply by astrelsky** [reply]
I actually didn't realize that was what `CodeUnit.getAddress` does. Good to know :)
